book/docs/1_introduction_to_data_engineering/1.1_what_is_data_engineering.md

```diff
@@ -0,0 +1,95 @@
+# 📌 1.1. 데이터 엔지니어링이란 무엇일까?
+
+---
+
+- 데이터 엔지니어링에 대한 정의
+- 데이터 엔지니어의 역할
+- 데이터 엔지니어의 핵심 역량
+
+## 1.1.1. 데이터 엔지니어링이란?
+
+> `데이터 엔지니어링`은 데이터를 대규모로 수집, 저장 및 분석하기 위한 시스템을 설계하고 구축하는 작업입니다.
+> 
+
+거의 모든 산업 분야에 적용되는 광범위한 분야입니다. 하나의 조직은 방대한 양의 데이터를 수집할 수 있는 능력이 필요하며 데이터 과학자와 분석가에게 데이터가 도달할 때까지 사용 가능한 상태가 되도록 `올바른 사람과 기술`이 필요합니다.
+
+뒤에서 다룰 ‘Data Engineering Lifecycle’을 미리 살펴보면서 데이터 엔지니어가 하는 일을 머릿속으로 그려보도록 하겠습니다.
+<br></br>
+
+📌 ***What Is the Data Engineering Lifecycle? - O’Reilly Medea. Fundamentals of Data Engineering***
+> `The data engineering lifecycle` comprises stages that turn raw data ingredients into a useful end product, ready for consumption by analysts, data scientists, ML engineers, and others.
+> 
+
+*→ 데이터 엔지니어링 수명 주기(Data Engineering LIfecycle)는 원시 데이터를 분석가, 데이터 과학자, ML 엔지니어 등이 사용할 수 있는 유용한 최종 제품으로 전환하는 단계로 구성됩니다.*
+
+<img src="images/1.1_data_engineering_life_cycle.png" width=100% height=100%>
```

Comment:
영전님 로컬에서 빌드해보니 아래와 같이 오류가 발생하고 있어서요..!
```bash
/Users/user/Documents/data-engineering-for-everybody/book/docs/1_introduction_to_data_engineering/1.0_about_bigdata.md:3: ERROR: Document or section may not begin with a transition.
/Users/user/Documents/data-engineering-for-everybody/book/docs/1_introduction_to_data_engineering/1.1_what_is_data_engineering.md:25: ERROR: Directive 'image': Invalid option value: (option: 'height'; value: 100%)
not a positive measure of one of the following units:
"em" "ex" "px" "in" "cm" "mm" "pt" "pc" ""
```

우선 아래와 같이 변경해 두겠습니다!
`![Untitled](./images/1.1_data_engineering_life_cycle.png)`

Reply:
@heehehe 리뷰 주셔서 감사합니다!
로컬에서 직접 빌드 해보고 테스트를 해봐야겠네요...!

Image를 화면에 꽉차게 조정하려 했는데 mark down상으로는 문제가 없으나 GitBook으로 빌드할 때는 문제가 된 것 같네요...!

희선님께서 수정해주신 코드로 로컬에서 빌드하고 다시 리뷰 요청드리면 될까요?!

Reply:
네 https://github.com/Pseudo-Lab/data-engineering-for-everybody/pull/18#pullrequestreview-1486554033 이 부분 로컬에서 문제 없으시면 다시 리뷰 요청 주시면 감사하겠습니다ㅎㅎ!
(이미지는 말씀주신 것처럼 jupyter book에서 `<img>` 태그로 하는걸 처리하지 못하는 모양입니다ㅠㅠ)

Reply:
@heehehe 
쉽지않네요 ㅎㅎ 
몇가지 큰 고비가 있었습니다 ㅎㅎ

## 수정 사항 공유
1. 브랜치 수정 과정에서 원격 저장소 업데이트
* Git Merge에서 충돌이 날 경우, 충돌을 해소하는데 어려움이 있었습니다.
* fast-forward로 해결되지 않아 rebase를 했는데... 맞는 방법인지 잘 모르겠네요...(* 의견 주시면 감사하겠습니다!)
2. Local 환경에서 Git Book Build
* Chapter의 Main Page(template_main_page.md or main_page.md)의 첫 줄이 H1 태그로 시작되지 않으면 빌드 이후에도 Git Book에서 챕터 전체 내용을 확인할 수 없었습니다

결과적으로 로컬 환경에서 빌드 이후 콘텐츠가 정상적으로 출력되는 것을 확인했습니다.
여러 컨트리뷰터가 작업을 하다보니 작업 도중에 버전이 바뀌는 경우가 있는데 협업을 위한 가이드를 개선시켜 나가면 좋겠다는 생각을 했습니다.

소중한 의견 주시면 감사하겠습니다 :)

Reply:
@heehehe 
쉽지않네요 ㅎㅎ
몇가지 큰 고비가 있었습니다 ㅎㅎ

수정 사항 공유
브랜치 수정 과정에서 원격 저장소 업데이트
Git Merge에서 충돌이 날 경우, 충돌을 해소하는데 어려움이 있었습니다.
fast-forward로 해결되지 않아 rebase를 했는데... 맞는 방법인지 잘 모르겠네요...(* 의견 주시면 감사하겠습니다!)
Local 환경에서 Git Book Build
Chapter의 Main Page(template_main_page.md or main_page.md)의 첫 줄이 H1 태그로 시작되지 않으면 빌드 이후에도 Git Book에서 챕터 전체 내용을 확인할 수 없었습니다
결과적으로 로컬 환경에서 빌드 이후 콘텐츠가 정상적으로 출력되는 것을 확인했습니다.
여러 컨트리뷰터가 작업을 하다보니 작업 도중에 버전이 바뀌는 경우가 있는데 협업을 위한 가이드를 개선시켜 나가면 좋겠다는 생각을 했습니다.

소중한 의견 주시면 감사하겠습니다 :)

Reply:
(늦은 시간 코멘트 죄송합니다🙇)
영전님 제쪽에서 다시 테스트 해보니 계속 https://github.com/Pseudo-Lab/data-engineering-for-everybody/pull/18#pullrequestreview-1486554033 오류가 발생하고 있어서요..!

일단 아래 오류는 https://github.com/Pseudo-Lab/data-engineering-for-everybody/commit/6deb2a4136e49924ac7ca3581fe07e269b6819c1 와 같이 처리해서 해결했는데요,
```
/Users/user/Documents/data-engineering-for-everybody/book/docs/1_introduction_to_data_engineering/1.0_about_bigdata.md:3: ERROR: Document or section may not begin with a transition.
/Users/user/Documents/data-engineering-for-everybody/book/docs/1_introduction_to_data_engineering/1.1_what_is_data_engineering.md:3: ERROR: Document or section may not begin with a transition.
/Users/user/Documents/data-engineering-for-everybody/book/docs/1_introduction_to_data_engineering/1.2_data_engineering_use_case.md:3: ERROR: Document or section may not begin with a transition.
/Users/user/Documents/data-engineering-for-everybody/book/docs/1_introduction_to_data_engineering/1.3_modern_data_engineering.md:3: ERROR: Document or section may not begin with a transition.
```

여전히 아래 오류는 나오고 있더라고요🥲
<details><summary>오류 펼쳐보기</summary>

```
% jupyter-book build --all book                                 
Running Jupyter-Book v0.15.1
Source Folder: /Users/user/Documents/data-engineering-for-everybody/book
Config Path: /Users/user/Documents/data-engineering-for-everybody/book/_config.yml
Output Path: /Users/user/Documents/data-engineering-for-everybody/book/_build/html
Running Sphinx v4.5.0
[etoc] Changing master_doc to 'intro'
myst v0.18.1: MdParserConfig(commonmark_only=False, gfm_only=False, enable_extensions=['html_image'], disable_syntax=[], all_links_external=False, url_schemes=['mailto', 'http', 'https'], ref_domains=None, highlight_code_blocks=True, number_code_blocks=[], title_to_header=False, heading_anchors=None, heading_slug_func=None, footnote_transition=True, words_per_minute=200, sub_delimiters=('{', '}'), linkify_fuzzy_links=True, dmath_allow_labels=True, dmath_allow_space=True, dmath_allow_digits=True, dmath_double_inline=False, update_mathjax=True, mathjax_classes='tex2jax_process|mathjax_process|math|output_area')
myst-nb v0.17.2: NbParserConfig(custom_formats={}, metadata_key='mystnb', cell_metadata_key='mystnb', kernel_rgx_aliases={}, execution_mode='off', execution_cache_path='', execution_excludepatterns=[], execution_timeout=30, execution_in_temp=False, execution_allow_errors=False, execution_raise_on_error=False, execution_show_tb=False, merge_streams=False, render_plugin='default', remove_code_source=False, remove_code_outputs=False, code_prompt_show='Show code cell {type}', code_prompt_hide='Hide code cell {type}', number_source_lines=False, output_stderr='show', render_text_lexer='myst-ansi', render_error_lexer='ipythontb', render_image_options={}, render_figure_options={}, render_markdown_format='commonmark', output_folder='build', append_css=True, metadata_to_fm=False)
Using jupyter-cache at: /Users/user/Documents/data-engineering-for-everybody/book/_build/.jupyter_cache
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 51 source files that are out of date
updating environment: [new config] 51 added, 0 changed, 0 removed
/Users/user/Documents/data-engineering-for-everybody/book/docs/0_checklist/list.md:1: WARNING: Document headings start at H2, not H1 [myst.header]                                                                     
/Users/user/Documents/data-engineering-for-everybody/book/docs/0_checklist/list.md:15: WARNING: Document headings start at H2, not H1 [myst.header]
/Users/user/Documents/data-engineering-for-everybody/book/docs/0_checklist/list.md:27: WARNING: Document headings start at H2, not H1 [myst.header]
/Users/user/Documents/data-engineering-for-everybody/book/docs/0_checklist/list.md:39: WARNING: Document headings start at H2, not H1 [myst.header]

Exception occurred:
  File "/Users/user/opt/anaconda3/lib/python3.9/site-packages/markdown_it/rules_block/state_block.py", line 137, in isEmpty
    return (self.bMarks[line] + self.tShift[line]) >= self.eMarks[line]
IndexError: list index out of range
The full traceback has been saved in /var/folders/fb/xtftd5y96nl3xc10kvb6cd4w0000gn/T/sphinx-err-o1pxmjpc.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
Traceback (most recent call last):
  File "/Users/user/opt/anaconda3/lib/python3.9/site-packages/jupyter_book/sphinx.py", line 171, in build_sphinx
    app.build(force_all, filenames)
  File "/Users/user/opt/anaconda3/lib/python3.9/site-packages/sphinx/application.py", line 330, in build
    self.builder.build_update()
  File "/Users/user/opt/anaconda3/lib/python3.9/site-packages/sphinx/builders/__init__.py", line 286, in build_update
    self.build(to_build,
  File "/Users/user/opt/anaconda3/lib/python3.9/site-packages/sphinx/builders/__init__.py", line 300, in build
    updated_docnames = set(self.read())
  File "/Users/user/opt/anaconda3/lib/python3.9/site-packages/sphinx/builders/__init__.py", line 407, in read
    self._read_serial(docnames)
  File "/Users/user/opt/anaconda3/lib/python3.9/site-packages/sphinx/builders/__init__.py", line 428, in _read_serial
    self.read_doc(docname)
  File "/Users/user/opt/anaconda3/lib/python3.9/site-packages/sphinx/builders/__init__.py", line 468, in read_doc
    doctree = read_doc(self.app, self.env, self.env.doc2path(docname))
  File "/Users/user/opt/anaconda3/lib/python3.9/site-packages/sphinx/io.py", line 181, in read_doc
    pub.publish()
  File "/Users/user/opt/anaconda3/lib/python3.9/site-packages/docutils/core.py", line 217, in publish
    self.document = self.reader.read(self.source, self.parser,
  File "/Users/user/opt/anaconda3/lib/python3.9/site-packages/sphinx/io.py", line 101, in read
    self.parse()
  File "/Users/user/opt/anaconda3/lib/python3.9/site-packages/docutils/readers/__init__.py", line 78, in parse
    self.parser.parse(self.input, document)
  File "/Users/user/opt/anaconda3/lib/python3.9/site-packages/myst_nb/sphinx_.py", line 86, in parse
    return super().parse(inputstring, document)
  File "/Users/user/opt/anaconda3/lib/python3.9/site-packages/myst_parser/parsers/sphinx_.py", line 69, in parse
    parser.render(inputstring)
  File "/Users/user/opt/anaconda3/lib/python3.9/site-packages/markdown_it/main.py", line 275, in render
    return self.renderer.render(self.parse(src, env), self.options, env)
  File "/Users/user/opt/anaconda3/lib/python3.9/site-packages/markdown_it/main.py", line 260, in parse
    self.core.process(state)
  File "/Users/user/opt/anaconda3/lib/python3.9/site-packages/markdown_it/parser_core.py", line 33, in process
    rule(state)
  File "/Users/user/opt/anaconda3/lib/python3.9/site-packages/markdown_it/rules_core/block.py", line 14, in block
    state.md.block.parse(
  File "/Users/user/opt/anaconda3/lib/python3.9/site-packages/markdown_it/parser_block.py", line 107, in parse
    self.tokenize(state, state.line, state.lineMax)
  File "/Users/user/opt/anaconda3/lib/python3.9/site-packages/markdown_it/parser_block.py", line 77, in tokenize
    if rule(state, line, endLine, False):
  File "/Users/user/opt/anaconda3/lib/python3.9/site-packages/markdown_it/rules_block/list.py", line 264, in list_block
    state.md.block.tokenize(state, startLine, endLine)
  File "/Users/user/opt/anaconda3/lib/python3.9/site-packages/markdown_it/parser_block.py", line 77, in tokenize
    if rule(state, line, endLine, False):
  File "/Users/user/opt/anaconda3/lib/python3.9/site-packages/markdown_it/rules_block/list.py", line 272, in list_block
    prevEmptyEnd = (state.line - startLine) > 1 and state.isEmpty(state.line - 1)
  File "/Users/user/opt/anaconda3/lib/python3.9/site-packages/markdown_it/rules_block/state_block.py", line 137, in isEmpty
    return (self.bMarks[line] + self.tShift[line]) >= self.eMarks[line]
IndexError: list index out of range

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/user/opt/anaconda3/bin/jupyter-book", line 8, in <module>
    sys.exit(main())
  File "/Users/user/opt/anaconda3/lib/python3.9/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/Users/user/opt/anaconda3/lib/python3.9/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/Users/user/opt/anaconda3/lib/python3.9/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/user/opt/anaconda3/lib/python3.9/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/user/opt/anaconda3/lib/python3.9/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/Users/user/opt/anaconda3/lib/python3.9/site-packages/jupyter_book/cli/main.py", line 317, in build
    builder_specific_actions(
  File "/Users/user/opt/anaconda3/lib/python3.9/site-packages/jupyter_book/cli/main.py", line 525, in builder_specific_actions
    raise RuntimeError(_message_box(msg, color="red", doprint=False)) from result
RuntimeError: 
===============================================================================

There was an error in building your book. Look above for the cause.

===============================================================================
```

</details>

제쪽에서는 세션 1 폴더를 다른 곳으로 옮기고 빌드해보면 정상적으로 동작해서
세션 1 파일 하나씩 추가해보면서 확인해보려고 하는데요,
혹시 영전님께서 빌드하실때는 위 오류 없으셨을까요..?

***

그리고 conflict 발생하는 경우는.. 아마 제가 commit 올린 부분과 따로 수정하신 부분이 겹치면서 발생했을 것 같은데요,
수정 진행하시기 전에 `git pull origin session-1` 하시면 괜찮지 않을까 싶습니다..!
혹시 위처럼 진행해도 충돌 발생하시면 말씀해주세요 :)